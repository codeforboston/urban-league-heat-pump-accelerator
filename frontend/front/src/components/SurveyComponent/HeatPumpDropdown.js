import { Controller, useController } from "react-hook-form";
import {
  FormControl,
  TextField,
  Select,
  MenuItem,
  InputLabel,
  FormLabel,
} from "@mui/material";
import React, { useCallback, useMemo } from "react";

/**
 * A radio group to be used with react-hook-form
 */
export const HeatPumpDropdown = ({
  name,
  control,
  options,
  label,
  disabled,
  required,
  enableOtherField,
  disableFancyLabel,
}) => {
  const Label = disableFancyLabel ? FormLabel : InputLabel;

  const { field: groupField, formState } = useController({ name, control });

  const otherFieldName = `${name}/other`;
  const showOtherInput = useMemo(
    () =>
      enableOtherField &&
      groupField.value &&
      typeof groupField.value == "string" &&
      groupField.value.toLowerCase() === "other",
    [enableOtherField, groupField]
  );

  const mainFieldError = formState.errors[name];
  const otherFieldError = formState.errors[otherFieldName];

  const renderDropdown = useCallback(
    (field) => {
      return (
        <FormControl fullWidth error={!!mainFieldError}>
          <Label id={`${name}-dropdown-label`} htmlFor={`${name}-dropdown`}>
            {label}
          </Label>
          <Select
            name={`${name}-dropdown`}
            aria-labelledby={`${name}-dropdown-label`}
            variant="filled"
            disabled={disabled}
            inputProps={{
              id: `${name}-dropdown`,
            }}
            {...field}
          >
            {options.map((option) => (
              <MenuItem value={option.value} key={`option-${option.value}`}>
                {option.label}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      );
    },
    [name, options, label, mainFieldError, disabled]
  );

  return (
    <>
      <Controller
        name={name}
        control={control}
        rules={
          required && {
            required: { value: true, message: "This field is required!" },
          }
        }
        render={({ field }) => renderDropdown(field)}
      />
      {showOtherInput && (
        <Controller
          name={otherFieldName}
          control={control}
          rules={
            required && {
              required: {
                value: showOtherInput,
                message: "This field is required!",
              },
            }
          }
          render={({ field }) => (
            <TextField
              label="Other..."
              variant="standard"
              {...field}
              error={!!otherFieldError}
              helperText={!!otherFieldError && otherFieldError.message}
              disabled={disabled}
            />
          )}
        ></Controller>
      )}
    </>
  );
};
