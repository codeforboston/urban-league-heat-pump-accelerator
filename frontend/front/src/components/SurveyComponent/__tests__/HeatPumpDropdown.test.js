import { fireEvent, render, screen, within } from "@testing-library/react";
import React from "react";
import { HeatPumpDropdown } from "../HeatPumpDropdown";
import { useForm } from "react-hook-form";

const YES_NO_FIELD = {
  id: 1,
  display_order: 1,
  text: "Have you heard of heat pumps?",
  response_type: "radio",
  response_options: ["Yes", "No"],
};

const OTHER_FIELD_LABEL = "What's your favorite Pokemon?";
const OTHER_FIELD = {
  id: 1,
  display_order: 1,
  text: OTHER_FIELD_LABEL,
  response_type: "radio",
  response_options: ["Bulbasaur", "Other"],
};

const DEFAULT_FORM = { 1: "" };

const FormWrapper = ({ q }) => {
  const { control } = useForm({ defaultValues: DEFAULT_FORM });
  return (
    <HeatPumpDropdown
      key={`q${q.id}`}
      control={control}
      name={`${q.id}`}
      label={q.text}
      options={q.response_options.map((o) => ({
        value: o,
        label: o,
      }))}
      {...q}
    />
  );
};

describe("HeatPumpDropdown", () => {
  it("renders as expected", () => {
    const { container } = render(<FormWrapper q={YES_NO_FIELD} />);

    expect(container).toMatchSnapshot();
  });

  it("enableOtherField=true shows other field when 'other' is selected", () => {
    render(<FormWrapper q={{ ...OTHER_FIELD, enableOtherField: true }} />);

    fireEvent.mouseDown(screen.getByRole("button"));
    const options = within(screen.getByRole("listbox"));
    fireEvent.click(options.getByText("Other"));

    expect(screen.getByLabelText("Other...")).not.toBeNull();
  });

  it("enableOtherField=false hides other field when 'other' is selected", () => {
    render(<FormWrapper q={OTHER_FIELD} />);

    fireEvent.mouseDown(screen.getByRole("button"));
    const options = within(screen.getByRole("listbox"));
    fireEvent.click(options.getByText("Other"));

    expect(screen.queryByLabelText("Other...")).toBeNull();
  });

  it("disableFancyLabel=true renders different label", () => {
    const { container } = render(
      <FormWrapper q={{ ...YES_NO_FIELD, disableFancyLabel: true }} />
    );

    expect(container).toMatchSnapshot();
  });
});
