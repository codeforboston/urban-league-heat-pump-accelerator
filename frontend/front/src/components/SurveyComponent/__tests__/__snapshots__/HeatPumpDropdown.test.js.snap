// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatPumpDropdown disableFancyLabel=true renders different label 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiFormLabel-colorPrimary css-u4tvz2-MuiFormLabel-root"
      for="1-dropdown"
      id="1-dropdown-label"
    >
      Have you heard of heat pumps?
    </label>
    <div
      aria-labelledby="1-dropdown-label"
      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl  css-llrb4p-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="mui-component-select-1"
        class="MuiSelect-select MuiSelect-filled MuiInputBase-input MuiFilledInput-input css-d9oaum-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
        id="mui-component-select-1"
        role="button"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        id="1-dropdown"
        name="1"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`HeatPumpDropdown renders as expected 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for="1-dropdown"
      id="1-dropdown-label"
    >
      Have you heard of heat pumps?
    </label>
    <div
      aria-labelledby="1-dropdown-label"
      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl  css-llrb4p-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="mui-component-select-1"
        class="MuiSelect-select MuiSelect-filled MuiInputBase-input MuiFilledInput-input css-d9oaum-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
        id="mui-component-select-1"
        role="button"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        id="1-dropdown"
        name="1"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
